Fix #4249: Add local storage support for collections played by guests.

BenHenning opened the pull request:
Adds support to the collection player to store local collection progress by guests.

**Checklist**
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

BenHenning committed:
Initial iteration of adding local storage support for collections played
by logged out users.

seanlip commented on core/templates/dev/head/domain/collection/GuestCollectionProgressService.js:
```diff
@@ -0,0 +1,146 @@
+
+oppia.factory('GuestCollectionProgressService', [
+  function() {
+    var COLLECTION_STORAGE_KEY = 'collectionProgressStore';
```
Should we add something like temp or v1 or some random string to the key, since this is meant to be temporary and we want to deprecate it later?

Also, might be worth having a "reset progress" button?

BenHenning replied:
Added v1, though there's no good notion of versioning here. If we want to clean this up/migrate it later, it should be simple to do.

A reset progress button is simple enough, but what's the benefit? I don't expect learners to accidentally complete an exploration, and they can always return to any exploration they wish to repeat. Do you think learners will want to experience the entire pathway again or just a single exploration?

seanlip replied:
Entire pathway. Useful if devices need to be reset after the trial.

seanlip replied:
(Suggest showing this only if there _is_ progress to be reset, btw.)

BenHenning replied:
So is the benefit for us to be able to reset the devices, or for users wanting to replay the collection in order? The latter seems less likely to me and is why I'm slightly reluctant to add this.

If the goal is to reset devices after the trial, we can just clear local storage on the devices themselves which will clear these temporary caches.

BenHenning replied:
Discussed offline; we're going to avoid doing this since it should be suitably easy for developers to clear local storage without a button, and we're instead going to whitelist the feature specifically to the fractions collection to avoid any user confusion that may come from saved state.

seanlip replied:
Just a note, let's have the whitelist be empty by default. We'll add the fractions collection id as part of the release process (I'll make a note).

Thanks!

(Oh and btw, there's another merge conflict, sorry)

seanlip commented on core/templates/dev/head/domain/collection/GuestCollectionProgressService.js:
```diff
@@ -0,0 +1,146 @@
+ * user's progress when they create an account.
+ */
+
+// TODO(bhenning): Move this to a shared service which stores state across the
```
Could you explain a bit more about what "shared" means in this context? (Isn't this already a singleton service?)

BenHenning replied:
Thanks for pointing this out; 'shared' was attributed to the wrong noun. What I meant is we should have a service that stores state which is shared across the frontend. I also provided a few examples for context. This is a lead-in for the saving exploration progress work.

seanlip commented on core/templates/dev/head/domain/collection/GuestCollectionProgressService.js:
```diff
@@ -0,0 +1,146 @@
+    var COLLECTION_STORAGE_KEY = 'collectionProgressStore';
+
+    var setCollectionProgressStore = function(store) {
+      window.localStorage[COLLECTION_STORAGE_KEY] = JSON.stringify(store);
```
Should use $window?

BenHenning replied:
Ah, it has been a while. Done!

seanlip commented on core/templates/dev/head/domain/collection/GuestCollectionProgressService.js:
```diff
@@ -0,0 +1,146 @@
+        setCollectionProgressStore(store);
+        return store;
+      }
+      return JSON.parse(store);
```
Should we be converting the stored thing to a domain object as soon as it's retrieved, rather than just working with raw dicts?

BenHenning replied:
Seems reasonable; done.

seanlip commented on core/templates/dev/head/domain/collection/GuestCollectionProgressService.js:
```diff
@@ -0,0 +1,146 @@
+       * Returns whether the guest user has made any progress toward completing
+       * the specified collection.
+       */
+      hasMadeProgress: function(collectionId) {
```
Perhaps hasCompletedSomeExploration is better

BenHenning replied:
Sure; done.

seanlip commented on core/templates/dev/head/pages/collection_player/CollectionPlayer.js:
```diff
@@ -254,13 +251,32 @@ oppia.controller('CollectionPlayer', [
+            CollectionPlaythroughObjectFactory.create(
+              nextExplorationIds, completedExplorationIds));
+        } else {
+          $scope.collectionPlaythrough = (
```
Is there a playthrough dict sent to the frontend in the case where the user isn't logged in and hasn't made any progress? Even if so, would it be better to just do the following?

```
   if (!$scope.isLoggedIn) {
     if (no progress) create default
     else use existing progress
   } else {
     use backend progress
   }
```

BenHenning replied:
Yes, per ``summary_services.get_learner_collection_dict_by_id`` we always pass a playthrough dict to the frontend, and specialize it in the logged out case to the initial exp IDs with no current progress. That actually is the 'default' your snippet above refers to, and I think it's easier to rely on the backend to provide the correct defaults rather than recreating them locally (though we could via CollectionObjectFactory's getStartingCollectionNodes() or by reconstructing them via the guest services methods, it's duplicating logic with the backend).

kevinlee12 commented on core/templates/dev/head/domain/collection/GuestCollectionProgressServiceSpec.js:
```diff
@@ -0,0 +1,205 @@
+// Copyright 2016 The Oppia Authors. All Rights Reserved.
```
Nit: year :)

BenHenning replied:
Done.

kevinlee12 reviewed (commented):
Not much to add onto Sean's comments, but it seems like `GuestCollectionProgressService` didn't get imported properly (perhaps others as well).

seanlip commented:
Oh, @BenHenning -- I just remembered something else. Shouldn't we whitelist this to only apply for the Fractions collection, for now?

anmolshkl commented:
hi @BenHenning, ping!

BenHenning committed:
First round of addressing review comments including introducing a new
GuestCollectionProgressObjectFactory and corresponding test rather than
relying on raw object access in the GuestCollectionProgressService.

BenHenning committed:
Address remaining reviews and fix broken protractor tests (exploration
editor requires the new guest collection progress service/object factory
dependencies).

BenHenning committed:
Fix linter error.

BenHenning commented:
Thanks!

Fixed the import issue in ``exploration_editor`` so e2e tests should pass now.

@seanlip does this seem like a reasonable long-term solution? I don't think it needs to be whitelisted for just the fractions lessons given that.

BenHenning committed:
Merge branch 'develop' into collection-guest-mode

Conflicts:
	PlayerServices.js

seanlip commented:
Hi @BenHenning -- did you see my earlier comment on resetting? I think that, as long as that's possible, that would allow people to undo any effects this may have, and then this PR can be considered a long-term solution.

seanlip commented:
Also please note the merge conflicts.

BenHenning commented:
Conflict is resolved and I left a reply to the earlier comment.

BenHenning committed:
Add whitelist for collection IDs to store guest progress.

BenHenning committed:
Merge branch 'develop' into collection-guest-mode

Conflicts:
	core/templates/dev/head/pages/exploration_player/PlayerServices.js

BenHenning commented:
Resolved new conflicts & added whitelist for gating the feature to specific collection IDs for the upcoming RCT. PTAL.

seanlip reviewed (approved):
Thanks @BenHenning!

BenHenning commented:
Thanks! Now waiting on Travis.
